### ISSUE TITLE
Disabling CTRL-P

### ISSUE THREAD

**Issue by BrendanMartin** [jupyterlab-contrib/jupyterlab-vim#21]
In insert mode, I sometimes hit CTRL-P when trying to hit CTRL-[ to leave insert mode. Unfortunately, Chrome launches the print dialog if you hit CTRL-P.

Right now I'm using this script as a bookmark to intercept it:
```
document.addEventListener('keydown', function(event) {
  if(event.ctrlKey && event.keyCode == 80) {
    event.preventDefault(); 
  }
});
```
But was wondering if this is something we could integrate into this extension.

**Comment by ianhi** [comment]
Your proposed fix has the downside of preventing of the print dialog from working, which if implemented would almost certainly create bug reports that this extension broke printing :upside_down_face: 

Admittedly using `ctrl-P` to just print the jlab interface is a pretty niche use case. But it still seems that a modification like this to prevent the occasional typo is outside the scope of this extension. In general unless a keyboard shortcut directly conflicts with a known `vim` keybinding then I don't think it's the role of this extension to mess with the browser's default behavior. 

Maybe something like https://chrome.google.com/webstore/detail/shortkeys-custom-keyboard/logpjaacgmcbpdkdchjiaagddngobkck would work in order to inject your js automatically?

**Comment by BrendanMartin** [comment]
My idea would be to allow the disabling of these kinds of browser shortcuts. So if someone uses the print shortcut often, they can just leave it on.

**Comment by manymachines** [comment]
I accidently hit `Shift-Escape` sometimes when I am already out of vim mode.  This opens the Chrome Task Manager and then one time I also went on to trigger End process by mistake too...

**Comment by fsondej** [comment]
I have a similar problem with ctrl+o in firefox. I agree that it would be invasive if this extension changed browser shortcuts, but at least we could document in README some way to this manually. @BrendanMartin could you give more details on how you run this script?

**Comment by ianhi** [comment]
Now that we have a real settings file I'm open to re-considering this as an opt-in setting. If anyone wants to have a go at PR for something like this it's ok to `@` me for a review

**Comment by firai** [comment]
You can now achieve this in your user preferences by adding following to the keyboard shortcut settings (Settings → Settings Editor → JSON Settings Editor → Keyboard Shortcuts). However, as mentioned above, I don't think blocking browser shortcuts should become a default for the extension itself unless it directly conflicts with a keyboard shortcut used by the extension.

Block <kbd>Ctrl+P</kbd> in Normal mode
```json
{
    "shortcuts": [
        Starts here
        {
            "command": "",
            "selector": ".jp-NotebookPanel[data-jp-vim-mode='true'] .jp-Notebook.jp-mod-editMode",
            "keys": ["Ctrl P"]
        }
        Ends here
    ]
}
```

Block <kbd>Ctrl+P</kbd> in Command mode
```json
{
    "shortcuts": [
        Starts here
        {
            "command": "",
            "selector": ".jp-Notebook.jp-mod-commandMode",
            "keys": ["Ctrl P"]
        }
        Ends here
    ]
}
```

**Comment by firai** [comment]
A summary of browser key bindings that have been suggested here to be blocked by the extension:

- <kbd>Ctrl+P</kbd> – I understand that the desire to block this binding is due to occasional typos, and it does not seem appropriate for the extension to block the browser shortcut by default due to this. However, a way to disable this in the user preferences is provided in https://github.com/jupyterlab-contrib/jupyterlab-vim/issues/21#issuecomment-1742018131.
- <kbd>Shift+Esc</kbd> – This browser key binding has been blocked by default in Jupyter Command mode since v4.02, and an option to disable the blocking key binding will be introduced in the next release (most likely to be named v4.04). Furthermore, the need to use <kbd>Shift+Esc</kbd> should reduce in the future, since <kbd>Esc</kbd> will be able to escape from vim Normal mode to Jupyter Command mode in the next version (again, most likely named v4.04).
- <kbd>Ctrl+O</kbd> – This key binding only applies in vim mode, and the key binding already auto-blocks the browser shortcut in this mode; repeated pressing of <kbd>Ctrl+O</kbd> in vim mode inserts cells above. If someone wishes to disable it in Jupyter Command mode for whatever reason, they can use the suggestion in https://github.com/jupyterlab-contrib/jupyterlab-vim/issues/21#issuecomment-1742018131 to do so.

Based on the above, I will close this issue as completed and pin this issue. If you would like to suggest another key binding that we should consider blocking, please open a new issue.
